### ISSUE TITLE
spirv-fuzz: Segmentation fault in FuzzerPassAddEquationInstructions

### ISSUE THREAD

**Issue by Vasniktel** [KhronosGroup/SPIRV-Tools#3216]
First of all, I know this issue is similar to #3213, but still, it's quite different from that case (at least it seems so). 
To reproduce:
- use this commit 2a20e6e
- compile the code
- `spirv-fuzz shader.spv -o fuzzed.spv --donors=donors.txt --seed=520311`

At this point, you will get a segfault.
What happens is:
- during one of the iterations `FuzzerPassAddEquationInstructions::Apply` collects instructions into `integer_instructions` variable here https://github.com/KhronosGroup/SPIRV-Tools/blob/da4cd214850296225484d3c2575c0e7eec83d607/source/fuzz/fuzzer_pass_add_equation_instructions.cpp#L76 The first instruction in that array will be an unsigned integer `OpConstant`.
- on this line https://github.com/KhronosGroup/SPIRV-Tools/blob/da4cd214850296225484d3c2575c0e7eec83d607/source/fuzz/fuzzer_pass_add_equation_instructions.cpp#L81 we select an instruction from the array at random, which in our case will be unsigned integer `OpConstantComposite` vector
- on this line https://github.com/KhronosGroup/SPIRV-Tools/blob/da4cd214850296225484d3c2575c0e7eec83d607/source/fuzz/fuzzer_pass_add_equation_instructions.cpp#L94 we get number of elements in the vector which will be `3` in our case
- then when we launch `RestrictToVectorWidth` function on this line https://github.com/KhronosGroup/SPIRV-Tools/blob/da4cd214850296225484d3c2575c0e7eec83d607/source/fuzz/fuzzer_pass_add_equation_instructions.cpp#L107 it will iterate over all instructions in `integer_instructions`. When we select the first instruction (which is `OpConstant`), we will get a segfault on this line https://github.com/KhronosGroup/SPIRV-Tools/blob/da4cd214850296225484d3c2575c0e7eec83d607/source/fuzz/fuzzer_pass_add_equation_instructions.cpp#L206 when `AsVector()` returns `nullptr`. (Remember that we computed `vector_width`  and set it to `3`).

All the necessary files are in [this gist](https://gist.github.com/Vasniktel/c5ecc0f0527be9b78f630389cb931e9c)

**Comment by afd** [comment]
Thanks for this!  Are you able to repro it (albeit with a different seed) on a master commit?  If so please list the commit here.  Issues that require seeds to reproduce can be painful because as the tool evolves, a seed that once revealed the problem ceases to do so.

**Comment by Vasniktel** [comment]
@afd, try 2a20e6e9 with seed `21041`.

**Comment by afd** [comment]
@Vasniktel A couple of questions on how to repro:

1. Should I use "after_applied_transformations.spv" as "shader.spv", or should I compile the GLSL shader using glslangValidator and use that as "shader.spv"?

2. Should donors.txt be an empty file or did you use some donor shaders?

**Comment by Vasniktel** [comment]
1. `after_applied_transformations.spv` is the same as `fuzzed.spv`. The reason I saved it is because when `spirv-fuzz` crashes, it doesn't leave any logs so I had to manually figure out what happened and print them. Compiling the shader with `glslangValidator` should be enough to reproduce the bug. Note that I didn't save the commit id when I created `after_applied_transformations.spv`, so when you try to run
```
spirv-fuzz shader.spv -o fuzzed.spv --donors=donors.txt --seed=520311
```
the sequence of transformations will be different but the reason for the crash remains the same.
2. `donnors.txt` is empty.
